perf-event/src/lib.rs

```diff
@@ -675,6 +825,117 @@ impl<'a> Builder<'a> {
     }
 }
 
+impl<'a> Builder<'a> {
+    /// Indicate additional values to include in the generated sample events.
+    /// Note that this method is additive and does not remove previously added
+    /// sample types.
+    ///
+    /// See the documentation of [`Sample`] for what's available to be
+    /// collected.
+    ///
+    /// # Example
+    /// Here we build a sampler that grabs the instruction pointer, process ID,
+    /// thread ID, and timestamp whenever the underlying event triggers a
+    /// sampling.
+    /// ```
+    /// # use perf_event::{Builder, Sample};
+    /// let mut sampler = Builder::new()
+    ///     .sample(Sample::IP)
+    ///     .sample(Sample::TID)
+    ///     .sample(Sample::TIME)
+    ///     .build_sampler(8192)?;
+    /// # Ok::<_, std::io::Error>(())
+    /// ```
+    pub fn sample(mut self, sample: Sample) -> Self {
+        self.attrs.sample_type |= sample.bits();
+        self
+    }
+
+    /// Enable the generation of [`Mmap`] records for `PROT_EXEC` mappings
+    /// when sampling. This allows tools to notice new executable code being
+    /// mapped into a program (e.g. dyamic shared libraries) so that addresses
+    /// can be mapped back to the original code.
+    ///
+    /// [`Mmap`]: crate::samples::Mmap
+    pub fn mmap(mut self, mmap: bool) -> Self {
+        self.attrs.set_mmap(mmap.into());
+        self
+    }
+
+    /// Set how many bytes will be written before an overflow notification
+    /// happens.
+    ///
+    /// Note only one of `wakup_watermark` and [`wakeup_events`] can be
+    /// configured.
+    ///
+    /// [`wakeup_events`]: Self::wakeup_events
+    pub fn wakeup_watermark(mut self, watermark: usize) -> Self {
+        self.attrs.set_watermark(1);
+        self.attrs.__bindgen_anon_2.wakeup_watermark = watermark as _;
+        self
+    }
+
+    /// Set how many samples will be written before an overflow notification
+    /// happens.
+    ///
+    /// Note that `wakeup_events` only counts [`RecordType::SAMPLE`] records.
+    /// To receive overflow notifications for all record type use
+    /// [`wakeup_watermark`] instead.
+    ///
+    /// Prior to Linux 3.0, setting `wakeup_events` to 0 resulted in no
+    /// overflow notifications; more recent kernels treat 0 the same as 1.
+    ///
+    /// [`RecordType::SAMPLE`]: crate::samples::RecordType::SAMPLE
+    /// [`wakeup_watermark`]: Self::wakeup_watermark
+    pub fn wakeup_events(mut self, events: usize) -> Self {
+        self.attrs.set_watermark(0);
+        self.attrs.__bindgen_anon_2.wakeup_events = events as _;
+        self
+    }
+
+    /// Construct a [`Sampler`] according to the specifications made on this
+    /// `Builder`. Requires that you specify a buffer size for the ring buffer.
+    /// This will be rounded up to the next power-of-two multiple of the page
+    /// size.
```

Comment:
Here and elsewhere: Generally the first paragraph of a doc string should be a one-line summary. This is what shows up in rustdoc search result lists, so it's the text that people are going to use to select a particular hit. Keeping it as a separate paragraph is a way to remember that this leading text has this important secondary role.

Reply:
I've fixed all the cases of this that I found. If you see another one point it out and I'll fix it.